src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -3321,6 +3328,26 @@ private void SetUpRPCParameters(_SqlRPC rpc, int startCount, bool inSchema, SqlP
                     // set default value bit
                     if (parameter.Direction != ParameterDirection.Output)
                     {
+                        // detect incorrectly derived type names unchanged by the caller and fix them
+                        if (parameter.IsDerivedParameterTypeName)
+                        {
+                            string[] parts = MultipartIdentifier.ParseMultipartIdentifier(parameter.TypeName, "[\"", "]\"", SR.SQL_TDSParserTableName, false);
+                            if (parts != null)
+                            {
+                                int length = parts.Length;
+                                if (
+                                    length >= 3 && //require at least 3 parts
+                                    parts[length - 1] != null && // name must not be null
+                                    parts[length - 2] != null && // schema must not be null
+                                    parts[length - 3] != null && // server should not be null or we don't need to remove it
```

Comment:
The comments are mismatched here I think, part[0] should be server name, part[1] is the database/catalog name.

Reply:
Note that they're `length-x` so the first reference is to the last element, so they're in the right order because it's `server.database.schema.name`. 
The whole multipart identifier parse is a bit awkward. I have a replacement written up that works in terms of spans but I'm still iterating on the api surface and making sure I have full coverage.

Reply:
Yeah you're right. What I meant is in this line `parts[length - 3] != null && // server should not be null or we don't need to remove it`, parts[length - 3] will be parts[1] which is the database name, but the comment says server -- `// server should not be null or we don't need to remove it`,

Reply:
I was overthinking it and adding complexity where it wasn't required. The overload used will always return an int[4] array and the results will be right justified in the array. I've simplified the checks which fixes this and the point raised below.